Improve: disable ocamlformat if no dot ocamlformat in the project

hhugo opened the pull request:
The PR adds a way to not reformat files that live in a "project" that doesn't contain a `.ocamlformat` file.

hhugo commented:
@samoht, @gpetiot, @jberdine, alternative suggestion welcome.

jberdine commented:
IIUC, there is a similar situation with dune, which has resulted in a way of specifying subdirs to ignore. This would result in putting a config file in the dir containing the project root dir to be ignored, and adding that dir's name into the config. I'm not sure if that is better, but think that the space was explored at some length for dune.

hhugo commented:
Just to be clear, the issue that this PR is trying to solve is about not formatting files that are not meant to be formatted. In particular, I don't want files from the ocaml compiler to be formatted when I edit them in emacs.

jberdine commented:
I see. I guess perhaps I should enable format-on-save myself since it seems most people use it that way.

For that particular use case, you prefer to call ocamlformat and have it detect that it's disabled, rather than e.g. configuring emacs to not call ocamlformat in the first place? I'm thinking of for instance enabling format-on-save in tuareg-mode-hook only if the file name matches some user-specific whitelist.

This PR basically says to look for an .ocamlformat file in the ancestor directories of the input file, but stop looking when reaching a "project root" dir, right? I'm not sure the implementation is the simplest, but the basic approach seems good to me. We should document it in the man page where it discusses where .ocamlformat files are searched for.

Perhaps this stop at project root mode should be the default?

hhugo commented:
I'm now curious about how you use ocamlformat if not with format-on-save. Some logic could be implemented in the emacs mode, but similar logic would then need to be implemented in other editor mode a well

samoht commented:
`M-q` only formats the current top-level expression, right? How do you do this without `--lines`? But maybe the expression is computed by tuareg mode already.

jberdine commented:
My binding in merlin-mode-map shadows tuareg's, so it formats the whole buffer.

hhugo commented:
I've reworked the PR.
The behaviour is now triggered by `--inside-project-only` (better name welcome).
This flags is now passed to ocamlformat in the emacs mode.
I've also added a `--print-config` flag that will print config updates together with their origin (see new tests)

hhugo commented:
I've reverted the change to the emacs mode for now so that the behaviour is preserved (to be discussed in another PR)

jberdine commented on src/Conf.ml:
```diff
@@ -726,6 +763,14 @@ let inputs =
+let inside_project_only =
+  let doc =
+    "Do not read .ocamlformat config files outside the project. If no \
+     config file is found, disable ocamlformat."
```
What is considered to be inside a project should be documented explicitly. Maybe: `The project root of an input file is taken to be the nearest ancestor directory that contains .git, .hg, or dune-project.`

hhugo replied:
updated

jberdine commented on src/Conf.ml:
```diff
@@ -253,30 +265,55 @@ end = struct
+  let log_update ~from ~name ~value =
+    match from with
+    | `Attribute -> ()
+    | (`File _ | `Config | `Commandline) as from ->
```
nit: `as from` redundant

hhugo replied:
it is not if you want the next match to be exhaustive.

jberdine replied:
Indeed, I missed that.

jberdine commented on src/Conf.ml:
```diff
@@ -726,6 +763,14 @@ let inputs =
+     config file is found, disable ocamlformat."
+  in
+  let default = false in
+  mk ~default Arg.(value & flag & info ["inside-project-only"] ~doc ~docs)
```
Is this an option that could productively use a separate env var?

hhugo replied:
I'm not sure about the env variable. I like your point about command line flags being more discoverable.

jberdine commented on src/Conf.ml:
```diff
@@ -993,20 +1055,50 @@ let kind_of fname =
+  in
+  let verbose = !print_config in
+  if verbose then
+    Option.iter project_root ~f:(Format.eprintf "project-root=%s\n%!") ;
```
call `to_relative`?

hhugo replied:
I don't think so.
It would be harder to read `project-root=../../../../`

hhugo commented on src/Conf.ml:
```diff
@@ -253,30 +265,55 @@ end = struct
+              let config = update config (parse value) in
+              if verbose then log_update ~from ~name ~value ;
+              Some (Ok config)
+            with _ -> Some (Error (`Bad_value (name, value)))
```
I've reduced the scope

jberdine commented on src/Conf.ml:
```diff
@@ -253,30 +265,55 @@ end = struct
+            match parse value with
+            | packed_value ->
+                if verbose then log_update ~from ~name ~value ;
+                Some (Ok (update config packed_value))
```
@hhugo I think it's ok to also push `update` out of the handler

hhugo commented:
I've implemented you suggestion `parse`, it now returns a result.

jberdine commented on src/Conf.ml:
```diff
@@ -764,6 +831,11 @@ let output =
       & info ["o"; "output"] ~doc ~docs ~docv)
 
+let print_config =
+  let doc = "Print config" in
```
nit: `.` at end.
